lib/streamlit/type_util.py

```diff
@@ -246,22 +256,46 @@ def is_function(x: object) -> TypeGuard[types.FunctionType]:
     return isinstance(x, types.FunctionType)
 
 
+def has_callable_attr(obj: object, name: str) -> bool:
+    """True if obj has the specified attribute that is callable."""
+    return hasattr(obj, name) and callable(getattr(obj, name))
+
+
 def is_namedtuple(x: object) -> TypeGuard[NamedTuple]:
-    t = type(x)
-    b = t.__bases__
-    if len(b) != 1 or b[0] is not tuple:
-        return False
-    f = getattr(t, "_fields", None)
-    if not isinstance(f, tuple):
-        return False
-    return all(type(n).__name__ == "str" for n in f)
+    """True if obj is an instance of a namedtuple."""
+    return isinstance(x, tuple) and has_callable_attr(x, "_asdict")
+
+
+def is_dataclass_instance(obj: object) -> bool:
+    """True if obj is an instance of a dataclass."""
+    return dataclasses.is_dataclass(obj) and not isinstance(obj, type)
```

Comment:
nit: could you add a small comment what the reason for the `isinstance(obj, type)` check is to shed some light on the context here.

Reply:
Added 👍

Reply:
great thank you!